### DISCUSSION TITLE
Jumpstart first budibase tool

### DISCUSSION THREAD

**Discussion by diedder** [Budibase/budibase#6125] [Help]
Hi,

I've found budisbase coincidentally and got the idea to use it for building a tool, which is in my mind for quite a long time. I'm a litte experienced in SQL und JS (via NodeRed), so maybe this lowcode tool can offer me the chance to realize it.

At them momemt I,ve only got a conceot I'd like to explain to you with the question, if this could be realizeable or it's to complex for the tool or my skills ;) The idea is about a webtool supporting me doing audit's. The preparation of checklists / questions an the documentaion.

There should be the following parts:
- a table of requirements, the audit criteria, parts of the standards
- a table of questions. these questions proove one ore more requirements
- for better preparation I'd like to collect multiple question in a checklist (template) . These collection can be used often for auditing the same topic in different audits
- when starting an audit, adding one audit to the audit table beside the organisational data I#d like to select one ore more of the checklist templates I'd like to perfom wihtin this audit. 
- And at least there must be some place to store the results of every answered question of the checklist.
![Unbenannt](https://user-images.githubusercontent.com/25912001/171099289-7fad88f8-9664-43ec-981d-4aaaa3f9844e.JPG)

Do you think this would be possible an a "good way" to do it? My goal would be to use best practise audit questions, collected by topics again to reduce preparation time.

My first practically question are
1) I'm looking for another way to manage relations than the relationspicker. The picker is not suitable for the hich amount of possible relations. For example: When adding a new checklist template I need to assign questions to it. there will be about 300 questions in the questiontable, so the small picker is not the best way. I would like to have a filterable list in the checklist-template-edit-page with all questions an beside every question a button to assign this question to the checklist or remove it. I hope I could explain my thougs of the UI.

2) What's best practise to generate a checklist of a checklist template to store the results? My first thought was to have an results-table. wher adding a checklist from template to an audit copy all questions of the checklisttemplae in die resultstable with some extra fields to store the result. Would this automation be possible?

I would be very happe to get some help or ideas für my projectvision from this communiy. Thank you in advance and sorry for my english...

br, diedder

**Comment by melohagan** [comment]
Hi @diedder 

I wasn't able to find anything "audit" specific, but it may be worth exploring some of the templates: https://budibase.com/templates/

With that said, I strongly recommend you utilise your SQL experience and use an external RDBMS instead of the internal table, as it seems this app will have strong reliance on relationships between tables.

I'm not a Data Architect, so I'm not going to review your ERD, but it does look reasonable to me and Budibase does have solid support for SQL databases and you should be able to accommodate a wide range of schemas. 

1. You could have a table with a button in each row (<a href="https://docs.budibase.com/docs/button#adding-buttons-to-table-rows">docs</a>) and a <a href="https://docs.budibase.com/docs/dynamic-filter">dynamic filter</a>.
    Alternatively, you can use a <a href="https://docs.budibase.com/docs/repeater">repeater</a> to display the question options, which would allow you greater control over the UI appearance.

2. I think you could use <a href="https://docs.budibase.com/docs/automation-actions#looping">automation looping</a> to accomplish this.

Hope this helps!

**Reply by diedder** [reply]
Hi @melohagan ,

Thank you for your answer! Following your advice I installed an external MySQL DB and connected Budibase to it. For my first test I'am using only 3 tables with relations (also defined in bubibase). I'm Feeding my two tables (see question 1 above) with two separate data providers which are using the defined SQL queries.

Until this point everthing works fine. I'm getting on two tables in the template page, the upper one is showing all question relatet to the template an the one below is showing all question without existent relation to the template. Both tables have also button in each row.
![image](https://user-images.githubusercontent.com/25912001/172636757-36e8b002-7471-4537-83b0-c1cfe1a30451.png)

But I didn't get the buttons working... I want the buttons to insert or delete a row in the relationstable, but SQL is not the issue. I want to use the ID of the question and/or the template in the SQL query but all my tries aren't working so far. It seems, that there is no access to the variables / data of the table/query.  Using the delete row action with the variable of the second column of the table as follows
![image](https://user-images.githubusercontent.com/25912001/172637882-a6ecfeaf-7f0a-4e5d-baa1-c31141dae8df.png)
isn't working and throws this error
![image](https://user-images.githubusercontent.com/25912001/172637680-208def92-f8be-47da-846d-65763d506209.png)

How can I access the data of the table or the above data provider for using it as ID or in my query? Please help...

br, diedder

**Reply by melohagan** [reply]
So based on the Delete Row screenshot, I would expect the following:

1. There is a table `Table IN` that has one-to-one relationship with table `transTemplateQuestion`
2. `getTemplateQuestions` is a column on table `Table IN`

Given that is the case, I think your binding should be: `{{ Table IN.getTemplateQuestions.0._id }}`

The `.0.` is added because relationships are always considered to be an array - even singleton relationships.
`_id` is used because it is always the row id. In the case of external tables, it will be the equivalent of the primary key. Even if you are using foreign keys in the relationship, you still need the primary key when doing CRUD operations on related data.

**Reply by diedder** [reply]
Update:
I found the issue inAN incorrect naming...
Thank's for your help!

**Reply by melohagan** [reply]
Brilliant, you're welcome!
